### PR TITLE
Class override

### PR DESCRIPTION
**Pull request by vojtakudlacek**
#41

### CODE REVIEW

File: runtime-decompiler/src/main/java/org/jrd/frontend/MainFrame/RewriteClassDialog.java [hunk 1]
```diff
@@ -48,23 +48,17 @@ private void setValidation()
...
         });
 
-        ok.addActionListener(new ActionListener() {
-            @Override
-            public void actionPerformed(ActionEvent e) {
-                ook[0] = true;
-                setVisible(false);
-            }
+        ok.addActionListener(e -> {
+            ook[0] = true;
```
Comment (judovana) [inline runtime-decompiler/src/main/java/org/jrd/frontend/MainFrame/RewriteClassDialog.java]: Although your commits made the original PR 100% better, the boolean[] is one of the things which must be fixed. Please get rid of it

File: decompiler_agent/src/main/java/org/jrd/agent/InstrumentationProvider.java [hunk 2]
```diff
@@ -27,10 +27,9 @@
         
     public void setClassBody(String cname, byte[] nwBody) throws UnmodifiableClassException {
         Class clazz = findClass(cname);
-        transformer.allowToSaveBytecode(); //?
+        transformer.allowToSaveBytecode();
         try {
             String nameWithSlashes = clazz.getName().replace(".", "/");
-            transformer.allowToSaveBytecode(); //? taken from getClassBody. Really twice?
```
Comment (judovana) [inline decompiler_agent/src/main/java/org/jrd/agent/InstrumentationProvider.java]: How had you verified this? Isnt the placement in try block better option?

File: runtime-decompiler/src/main/java/org/jrd/frontend/MainFrame/RewriteClassDialog.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        buttons.add(nothing);
+        buttons.add(ok);
+        this.add(inputs);
+        this.add(buttons, BorderLayout.EAST);
+        this.pack();
+    }
+
+    public boolean[] getOok()
+    {
+        return ook;
```
Comment (judovana) [inline runtime-decompiler/src/main/java/org/jrd/frontend/MainFrame/RewriteClassDialog.java]: The accursed booelan[] is also here. Do you know why it was here originally? (if not, ask, to learn what to never do)/ Please get rid of it.

File: decompiler_agent/src/main/java/org/jrd/agent/InstrumentationProvider.java (hunk 2 above)
Comment (vojtakudlacek) [reply decompiler_agent/src/main/java/org/jrd/agent/InstrumentationProvider.java]: Yes, I was looking down the function and it's only setting variable to true. And I tried it too.

Comment (vojtakudlacek) [reply decompiler_agent/src/main/java/org/jrd/agent/InstrumentationProvider.java]: If it is in the try block it is't working

Comment (judovana) [reply decompiler_agent/src/main/java/org/jrd/agent/InstrumentationProvider.java]: Interesting. Thank you!

**Comment by judovana** [comment]
I do not have much against. Looking forward for Radek's sharp eye.

**Comment by RadekManak** [comment]
I wouldn't be that optimistic. How exactly is it supposed to work? First I tried to create a dummy application that prints a string in a loop. Then I figured out I need to save modified java file. And compile it?

After reading some code, I figured out I need it in specific directory path. Finally, a green text! Let's press OK! Class rewrite failed. I gave up after 30 minutes.

**Comment by judovana** [comment]
First issue - yes, you are currently responsible for providing byte code you got on your own.
Already this have an benefit - imagine  a long time running application which suddenly started to throw errors.  And you own sources. So you make changes, build, and iunstall new class.

Second issue - class overwriting is very very limited. Eg you can not chage method signatures or add/remove them. Web search will tell you more (note, I searched right now and failed to find specification, just shards). it would be nice to implement mechanism how to get errors from agent. In this case it would be extra useful. Sory for wasting 30 minutes - try in jrd decompiling itself, eg replace cli class. that worked for me (rigth now)

Third issue - to allow the JRD to chnage class, compile it and return (where return is already implemented in this PR)  will be future work. The compilation itself is easy, but to obtain depndencies resolved to surprisingly tricky issue. If you provide an CP, then it is again easy. But to obtain it from running VM  had surprisingly a lot of challenges.

**Comment by judovana** [comment]
Oh btw - the patch is targeting also agent. Could it hapen you had system agent used instead of freshly built one?

**Comment by judovana** [comment]
Hi! Pushing. Radek, If you find any more issues,dont hesitate to let me know. Will fix it.

**Pull request merged**